File: src/pkce.zig
const std = @import("std");
const crypto = std.crypto;
const base64 = std.base64;

/// PKCE (Proof Key for Code Exchange) implementation
/// RFC 7636: https://tools.ietf.org/html/rfc7636
pub const Pkce = struct {
    code_verifier: [43]u8,
    code_challenge: [43]u8,

    /// Generate a new PKCE challenge pair
    pub fn generate(allocator: std.mem.Allocator) !Pkce {
        _ = allocator;

        var pkce: Pkce = undefined;

        // Generate 32 random bytes for code_verifier
        var random_bytes: [32]u8 = undefined;
        crypto.random.bytes(&random_bytes);

        // Base64 URL encode without padding
        const encoder = base64.url_safe_no_pad.Encoder;
        _ = encoder.encode(&pkce.code_verifier, &random_bytes);

        // Create SHA256 hash of code_verifier
        var hash: [32]u8 = undefined;
        crypto.hash.sha2.Sha256.hash(&pkce.code_verifier, &hash, .{});

        // Base64 URL encode the hash for code_challenge
        _ = encoder.encode(&pkce.code_challenge, &hash);

        return pkce;
    }

    /// Get the code verifier
    pub fn getCodeVerifier(self: *const Pkce) []const u8 {
        return &self.code_verifier;
    }

    /// Get the code challenge
    pub fn getCodeChallenge(self: *const Pkce) []const u8 {
        return &self.code_challenge;
    }

    /// Get the code challenge method (always S256)
    pub fn getCodeChallengeMethod() []const u8 {
        return "S256";
    }
};

test "PKCE generation" {
    const allocator = std.testing.allocator;

    const pkce_pair = try Pkce.generate(allocator);

    // Verify lengths
    try std.testing.expectEqual(43, pkce_pair.code_verifier.len);
    try std.testing.expectEqual(43, pkce_pair.code_challenge.len);

    // Verify they are different
    try std.testing.expect(!std.mem.eql(u8, &pkce_pair.code_verifier, &pkce_pair.code_challenge));
}

test "PKCE generates different values each time" {
    const allocator = std.testing.allocator;

    const pkce1 = try Pkce.generate(allocator);
    const pkce2 = try Pkce.generate(allocator);

    try std.testing.expect(!std.mem.eql(u8, &pkce1.code_verifier, &pkce2.code_verifier));
    try std.testing.expect(!std.mem.eql(u8, &pkce1.code_challenge, &pkce2.code_challenge));
}

File: src/pkce.rs
/// PKCE (Proof Key for Code Exchange) implementation
/// RFC 7636: https://tools.ietf.org/html/rfc7636
use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine};
use rand::Rng;
use sha2::{Digest, Sha256};

/// PKCE challenge pair containing code verifier and code challenge
#[derive(Debug, Clone)]
pub struct Pkce {
    code_verifier: String,
    code_challenge: String,
}

impl Pkce {
    /// Generate a new PKCE challenge pair
    ///
    /// Creates a cryptographically secure random code verifier and derives
    /// the code challenge using SHA256.
    ///
    /// # Examples
    ///
    /// ```
    /// use schlussel::pkce::Pkce;
    ///
    /// let pkce = Pkce::generate();
    /// assert_eq!(Pkce::code_challenge_method(), "S256");
    /// ```
    pub fn generate() -> Self {
        // Generate 32 random bytes for code_verifier
        let mut rng = rand::thread_rng();
        let random_bytes: [u8; 32] = rng.gen();

        // Base64 URL encode without padding
        let code_verifier = URL_SAFE_NO_PAD.encode(random_bytes);

        // Create SHA256 hash of code_verifier
        let mut hasher = Sha256::new();
        hasher.update(code_verifier.as_bytes());
        let hash = hasher.finalize();

        // Base64 URL encode the hash for code_challenge
        let code_challenge = URL_SAFE_NO_PAD.encode(hash);

        Self {
            code_verifier,
            code_challenge,
        }
    }

    /// Get the code verifier
    pub fn code_verifier(&self) -> &str {
        &self.code_verifier
    }

    /// Get the code challenge
    pub fn code_challenge(&self) -> &str {
        &self.code_challenge
    }

    /// Get the code challenge method (always S256)
    pub fn code_challenge_method() -> &'static str {
        "S256"
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_pkce_generation() {
        let pkce = Pkce::generate();

        // Verify lengths (base64 encoded 32 bytes = 43 chars without padding)
        assert_eq!(pkce.code_verifier().len(), 43);
        assert_eq!(pkce.code_challenge().len(), 43);

        // Verify they are different
        assert_ne!(pkce.code_verifier(), pkce.code_challenge());
    }

    #[test]
    fn test_pkce_generates_different_values() {
        let pkce1 = Pkce::generate();
        let pkce2 = Pkce::generate();

        assert_ne!(pkce1.code_verifier(), pkce2.code_verifier());
        assert_ne!(pkce1.code_challenge(), pkce2.code_challenge());
    }

    #[test]
    fn test_code_challenge_method() {
        assert_eq!(Pkce::code_challenge_method(), "S256");
    }
}
